contracts/poseidon_host_bench/lib.rs

```diff
@@ -0,0 +1,65 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+#![feature(min_specialization)]
+
+#[ink::contract(env = baby_liminal_extension::BabyLiminalEnvironment)]
+mod poseidon_host_bench {
+    use ink::storage::Mapping;
+    use openbrush::traits::Storage;
+    type MerkleHash = [u64; 4];
+
+    #[ink(storage)]
+    #[derive(Default, Storage)]
+    pub struct PoseidonHostBench {
+        hashes: Mapping<u32, MerkleHash>,
+        next_free_leaf: u32,
+    }
+
+    impl PoseidonHostBench {
+        #[ink(constructor)]
+        pub fn new(max_leaves: u32) -> Self {
+            if !max_leaves.is_power_of_two() {
+                panic!("Please have 2^n leaves")
+            }
+
+            let mut bench = Self::default();
+            bench.next_free_leaf = max_leaves;
+
+            bench.create_new_leaf([1, 7, 2, 9]);
+
+            bench
+        }
+
+        #[ink(message)]
+        pub fn hash(&mut self) {
+            self.hash_two([2, 1, 3, 7], [1, 7, 2, 9]);
+        }
+
+        #[ink(message)]
+        pub fn generate_path(&mut self) {
+            self.create_new_leaf([2, 1, 3, 7])
+        }
+
+        fn tree_value(&self, idx: u32) -> MerkleHash {
+            self.hashes.get(idx).unwrap_or_default()
+        }
+
+        fn create_new_leaf(&mut self, value: MerkleHash) {
+            self.hashes.insert(self.next_free_leaf, &value);
+
+            let mut parent = self.next_free_leaf / 2;
+            while parent > 0 {
+                let left_child = self.tree_value(2 * parent);
+                let right_child = self.tree_value(2 * parent + 1);
+                let parent_hash = self.hash_two(left_child, right_child);
+                self.hashes.insert(parent, &parent_hash);
+                parent /= 2;
+            }
+
+            self.next_free_leaf += 1;
+        }
+
+        fn hash_two(&self, x: MerkleHash, y: MerkleHash) -> MerkleHash {
```

Comment:
why not call it `two_to_one_hash` following the same convention

Reply:
done